File: compiler/engine/testdata/reference/memory_access/engine.go
package p

type N struct {
	Beat  float64 `sonolus:"imported"`
	Value float64 `sonolus:"memory"`
	Saved float64 `sonolus:"memory"`
}

func (n N) Initialize() {
	n.Value = 0
	n.Saved = 0
}

func (n N) UpdateSequential() {
	n.Value = n.Beat * 2
	if n.Value > 100 {
		n.Saved = n.Value
		n.Value = 0
	}
}

File: compiler/engine/testdata/reference/memory_access/engine.ts
// TypeScript equivalent of memory_access/engine.go.
import { Archetype, entityMemory, defineImport } from '@sonolus/sonolus.js-compiler/play'

const archetype = new Archetype('N')
const beat = defineImport('Beat')
const value = entityMemory(Number)
const saved = entityMemory(Number)
defineImport(value)
defineImport(saved)

export function initialize() {
	value.set(0)
	saved.set(0)
}

export function updateSequential() {
	value.set(beat * 2)
	if (value.get() > 100) {
		saved.set(value.get())
		value.set(0)
	}
}
